### ISSUE TITLE
Resuming from checkpoint doesn't seem to work

### ISSUE THREAD

**Issue by Superskyyy** [huggingface/trl#2657]
### Reproduction

I've been using PPO/RLOOTrainer and it seems the resume_from_checkpoint doens't work, upon seeing the code I was surprised that nothing implemented the checkpointing loading mechanism not even using the one from huggingface transfomers, (the trainer.train method doesn't take a resume_from_checkpoint arg)

How can I load back the checkpoint and resume the training? I assume people have been using this feature and I somehow missed the guide to do so.

### System Info

- Platform: Linux-5.15.0-124-generic-x86_64-with-glibc2.36
- Python version: 3.12.8
- PyTorch version: 2.5.1
- CUDA device(s): NVIDIA H100 80GB HBM3, NVIDIA H100 80GB HBM3, NVIDIA H100 80GB HBM3, NVIDIA H100 80GB HBM3, NVIDIA H100 80GB HBM3, NVIDIA H100 80GB HBM3, NVIDIA H100 80GB HBM3, NVIDIA H100 80GB HBM3
- Transformers version: 4.48.0
- Accelerate version: 0.34.2
- Accelerate config: 
  - compute_environment: LOCAL_MACHINE
  - distributed_type: DEEPSPEED
  - use_cpu: False
  - debug: False
  - num_processes: 8
  - machine_rank: 0
  - num_machines: 1
  - rdzv_backend: static
  - same_network: True
  - main_training_function: main
  - enable_cpu_affinity: False
  - deepspeed_config: {'deepspeed_config_file': 'deepspeed_config.json', 'zero3_init_flag': False}
  - downcast_bf16: no
  - tpu_use_cluster: False
  - tpu_use_sudo: False
  - tpu_env: []
- Datasets version: 3.2.0
- HF Hub version: 0.27.1
- TRL version: 0.14.0.dev0
- bitsandbytes version: not installed
- DeepSpeed version: 0.15.4
- Diffusers version: not installed
- Liger-Kernel version: not installed
- LLM-Blender version: not installed
- OpenAI version: not installed
- PEFT version: not installed

### Checklist

- [x] I have checked that my issue isn't already filed (see [open issues](https://github.com/huggingface/trl/issues?q=is%3Aissue))
- [x] I have included my system information
- [x] Any code provided is minimal, complete, and reproducible ([more on MREs](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any code provided is properly formatted in code blocks, (no screenshot, [more on code blocks](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any traceback provided is complete

**Comment by Superskyyy** [comment]
@qgallouedec Please give some insight to this as this is blocking my training, if it needs some implementation i will contribute back to the repo. Thanks!

**Comment by Superskyyy** [comment]
I have a very rough implementation following the transformers trainers design and seems to be working. Though I'm not sure why seems like all the trainers in TRL didn't support resuming.

**Comment by ShikunLi** [comment]
@Superskyyy  hello, could you share your  implementation of resuming from checkpoint? I really need this feature and it will help me a lot ! Thanks !

**Comment by qgallouedec** [comment]
@ShikunLi Are you getting an error when you try to resume from checkpoint? It is supposed to work

**Comment by Amrita-Bhattacherjee** [comment]
@qgallouedec 
There is no error, but the training starts from scratch, even when passing the checkpoint path including relevant trainer/optimizer states to `resume_from_checkpoint` in TrainingArguments

**Comment by indicava** [comment]
> [@ShikunLi](https://github.com/ShikunLi) Are you getting an error when you try to resume from checkpoint? It is supposed to work

Looking at the `ppo_trainer.py` code:

it unconditionally initializes to "0"
```
# trainer state initialization
        self.state.global_step = 0
        self.state.episode = 0
        self.state.max_steps = args.num_total_batches
        self.state.num_train_epochs = args.total_episodes / self.train_dataset_len
```

also, training loop always seems to start from 1
```
for update in range(1, args.num_total_batches + 1):
```

Am I missing something?

**Comment by qgallouedec** [comment]
PPO needs heavy refactored, so the issue probably still exists.
For RLOO, it's fixed with #3801

**Comment by RaccoonOnion** [comment]
trl.SFTTrainer has issue with restarting from checkpoint as well. 

The issue is similar to what @qgallouedec describe above:

> [@qgallouedec](https://github.com/qgallouedec) There is no error, but the training starts from scratch, even when passing the checkpoint path including relevant trainer/optimizer states to `resume_from_checkpoint` in TrainingArguments

It would be really helpful if this issue could be resolved since a lot of users do not have long, continuous session for them to finish training in one go.

**Comment by DaoyuanLi2816** [comment]
Following up specifically on the `SFTTrainer` report from @RaccoonOnion, since it hadn't been investigated yet (the rest of this thread is about PPO/RLOO, which @qgallouedec already addressed above: PPO needs a heavier refactor, RLOO's resume was fixed by #3801).

`SFTTrainer` doesn't override `train()` or anything checkpoint-related — it fully delegates to `transformers.Trainer`, which does support `resume_from_checkpoint` out of the box. I tried to reproduce the "starts from scratch" symptom on current `main` with a minimal single-process, no-DeepSpeed setup:

```python
args = SFTConfig(output_dir=out_dir, max_steps=4, per_device_train_batch_size=4,
                  save_strategy="steps", save_steps=2, ...)
trainer = SFTTrainer(model=..., args=args, train_dataset=ds, processing_class=tok)
trainer.train()  # runs steps 1-4, checkpoint-2 saved after step 2

# fresh model + fresh trainer
trainer2 = SFTTrainer(model=..., args=args, train_dataset=ds, processing_class=tok)
trainer2.train(resume_from_checkpoint="checkpoint-2")
```

and couldn't reproduce it: the resumed run's LR-scheduler values for steps 3-4 (`1e-05`, `5e-06`) exactly match the original run's steps 3-4, rather than restarting from step 1's initial LR (`2e-05`), and the resumed run's wall-clock time is ~half the original (2 steps actually executed, not 4) — both consistent with a correct resume, not a restart.

So the base case looks fine on current `main`. If you're still seeing this, it'd help to know: which trl/transformers version, and are you using DeepSpeed/FSDP or multi-GPU? Given the original issue was DeepSpeed-specific, it's possible what you hit is a distributed-state-restoration gap rather than something in `SFTTrainer` itself, and it'd need a repro under that setup to pin down.

**Comment by qgallouedec** [comment]
Thanks, I'm going to close this issue because it's solved for all trainers but PPO, and PPO is in maintenance mode, ie we won't fix/add features anymore to it.
